src/lib.rs

```diff
@@ -64,6 +54,162 @@ pub struct LoggingConnection<C: Connection> {
+
+            let result = self.conn.execute(query);
+            let duration = start_time.elapsed();
+
+            log_query(query, duration, time_utc, self.log_mode);
+            result
+        }
+    }
+
+    fn query_by_index<T, U>(&self, source: T) -> QueryResult<Vec<U>>
+    where
+        T: AsQuery,
+        T::Query: QueryFragment<Self::Backend> + QueryId,
+        Self::Backend: HasSqlType<T::SqlType>,
+        U: Queryable<T::SqlType, Self::Backend>,
+    {
+        let query = source.as_query();
+
+        if self.log_mode.do_not_log() {
+            self.conn.query_by_index(query)
+        } else {
+            let debug_query = diesel::debug_query::<Self::Backend, _>(&query).to_string();
+
+            let time_utc = chrono::Utc::now();
+            let start_time = std::time::Instant::now();
+
+            let result = self.conn.query_by_index(query);
+            let duration = start_time.elapsed();
+
+            log_query(&debug_query, duration, time_utc, self.log_mode);
+            result
+        }
+    }
+
+    fn query_by_name<T, U>(&self, source: &T) -> QueryResult<Vec<U>>
+    where
+        T: QueryFragment<Self::Backend> + QueryId,
+        U: QueryableByName<Self::Backend>,
+    {
+        if self.log_mode.do_not_log() {
+            self.conn.query_by_name(source)
+        } else {
+            let debug_query = diesel::debug_query::<Self::Backend, _>(&source).to_string();
+
+            let time_utc = chrono::Utc::now();
+            let start_time = std::time::Instant::now();
+            let result = self.conn.query_by_name(source);
+            let duration = start_time.elapsed();
+
+            log_query(&debug_query, duration, time_utc, self.log_mode);
+            result
+        }
+    }
+
+    fn execute_returning_count<T>(&self, source: &T) -> QueryResult<usize>
+    where
+        T: QueryFragment<Self::Backend> + QueryId,
+    {
+        if self.log_mode.do_not_log() {
+            self.conn.execute_returning_count(source)
+        } else {
+            let debug_query = diesel::debug_query::<Self::Backend, _>(&source).to_string();
+
+            let time_utc = chrono::Utc::now();
+            let start_time = std::time::Instant::now();
+
+            let result = self.conn.execute_returning_count(source);
+            let duration = start_time.elapsed();
+
+            log_query(&debug_query, duration, time_utc, self.log_mode);
+            result
+        }
+    }
+
+    fn transaction_manager(&self) -> &Self::TransactionManager {
+        // this is actually fine because we have an #[repr(transparent)]
+        // on LoggingTransactionManager, which means the layout is the same
+        // as the inner type
+        // See the ref-cast crate for a longer version: https://github.com/dtolnay/ref-cast
+        unsafe {
+            &*(self.conn.transaction_manager() as *const _ as *const Self::TransactionManager)
+        }
+    }
+
+    fn begin_test_transaction(&self) -> QueryResult<()> {
```

Comment:
The original [diesel code here](https://github.com/diesel-rs/diesel/blob/fa826f0c97e1f47eef34f37cb5b60056855a2b9a/diesel/src/connection/mod.rs#L116) is:
```rust
        let transaction_manager = self.transaction_manager();
        assert_eq!(transaction_manager.get_transaction_depth(), 0);
        transaction_manager.begin_transaction(self)
```

I do not see why this should behave different to this updated version if the statement in the comment above the unsafe block in `transaction_manager` remains true. If that's an false assumption we have a much larger problem than `begin_test_transaction`, because that would mean that we call into the wrong? `TransactionManager` in some places, which would ultimately mean that we need to consider transactions generally broken in diesel-oci

Reply:
Just to have the answer here as well:
> diesel-oci overwrites the default implementation: https://github.com/GiGainfosystems/diesel-oci/blob/fix_binary_double_byte_order_giga_head_cherries/src/oracle/connection/mod.rs#L79

Reply:
I do have some evidence that at least for test transactions we do have a total mess in `diesel-oci`. Eg, run a random test, and then run the same test again, and it will fail, because the input from the first one has not been deleted (even before we start using `diesel_logger`, so this may be worth looking at. However, that is not a guarantee that all transactions are broken for oracle.

I think you would have to properly test this with a real database running with `diesel-oci` by doing stuff, crashing the database and then trying to roll back. That would likely give a closer-to-definitive answer. My  understanding of Oracle only goes as far as:

  "Oracle is really complicated, and you need to be very experienced with Oracle databases to know if and how you've screwed up."

So I am not really qualified to give suggestions. But I do think without proper stress testing of an application that uses `diesel-oci`, there will not be a good answer.